checker/cost.go

```diff
@@ -0,0 +1,536 @@
+		switch ck := ek.ConstExpr.ConstantKind.(type) {
+		case *exprpb.Constant_StringValue:
+			v = int64(len(ck.StringValue))
+		case *exprpb.Constant_BytesValue:
+			v = int64(len(ck.BytesValue))
+		default:
+			return nil
+		}
+	case *exprpb.Expr_ListExpr:
+		v = int64(len(ek.ListExpr.Elements))
+	case *exprpb.Expr_StructExpr:
+		if ek.StructExpr.MessageName == "" {
+			v = int64(len(ek.StructExpr.Entries))
+		}
+	default:
+		return nil
+	}
+
+	return &v
+}
+
+var (
+	identCost  = CostEstimate{Min: 1, Max: 1}
+	selectCost = CostEstimate{Min: 1, Max: 1}
+	constCost  = CostEstimate{Min: 0, Max: 0}
+)
+
+// CostEstimate represents an estimated cost range.
+type CostEstimate struct {
+	Min, Max int64
+}
+
+// Add adds the costs and returns the sum.
+func (ce CostEstimate) Add(rhs CostEstimate) CostEstimate {
+	return CostEstimate{
+		ce.Min + rhs.Min,
+		ce.Max + rhs.Max,
+	}
+}
+
+type coster struct {
+	env                *Env
+	mappings           *mapping
+	freeTypeVarCounter int
+	sourceInfo         *exprpb.SourceInfo
+	types              map[int64]*exprpb.Type
+	references         map[int64]*exprpb.Reference
+	estimator          CostEstimator
+}
+
+// Cost estimates the cost of the parsed CEL expression.
+// The parsedExpr is expected to have passed type checking. If it has not been
+// type checked, the results are undefined.
+func Cost(parsedExpr *exprpb.ParsedExpr,
+	env *Env, estimator CostEstimator) CostEstimate {
+	c := coster{
+		env:        env,
+		mappings:   newMapping(),
+		sourceInfo: parsedExpr.GetSourceInfo(),
+		types:      make(map[int64]*exprpb.Type),
+		references: make(map[int64]*exprpb.Reference),
+		estimator:  estimator,
+	}
+	return c.cost(parsedExpr.GetExpr())
+}
+
+func (c *coster) cost(e *exprpb.Expr) CostEstimate {
+	if e == nil {
+		return CostEstimate{}
+	}
+
+	var cost CostEstimate
+	switch e.ExprKind.(type) {
+	case *exprpb.Expr_ConstExpr:
+		cost = constCost
+	case *exprpb.Expr_IdentExpr:
+		cost = c.costIdent(e)
+	case *exprpb.Expr_SelectExpr:
+		cost = c.costSelect(e)
+	case *exprpb.Expr_CallExpr:
+		cost = c.costCall(e)
+	case *exprpb.Expr_ListExpr:
+		cost = c.costCreateList(e)
+	case *exprpb.Expr_StructExpr:
+		cost = c.costCreateStruct(e)
+	case *exprpb.Expr_ComprehensionExpr:
+		cost = c.costComprehension(e)
+	default:
+		return CostEstimate{}
+	}
+	return cost
+}
+
+func (c *coster) costIdent(e *exprpb.Expr) CostEstimate {
+	identExpr := e.GetIdentExpr()
+	// Check to see if the identifier is declared.
+	if ident := c.env.LookupIdent(identExpr.GetName()); ident != nil {
+		c.setType(e, ident.GetIdent().Type)
+		// Overwrite the identifier with its fully qualified name.
+		identExpr.Name = ident.GetName()
+		return identCost
+	}
+
+	// undeclared reference
+	c.setType(e, decls.Error)
+	return CostEstimate{}
+}
+
+func (c *coster) costSelect(e *exprpb.Expr) CostEstimate {
+	sel := e.GetSelectExpr()
+	var sum CostEstimate
+	// Before traversing down the tree, try to interpret as qualified name.
```

Comment:
I would refrain from copying the checker code here. All of the information you want should be in the `references` and `types` derived from the `CheckedExpr.reference_map` and `CheckedExpr.type_map` respectively.

Reply:
Ahh, this is what I needed. The duplication was really bothering me. This trims the code up substantially.